Lazy methods: allow driver program to provide stubs

smowton opened the pull request:
This adds a mechanism for asking languaget instances to produce functions they have not advertised.
DeepTest currently uses it to supply stub function bodies (jbmc replaces the calls, but this results
in a less-useful trace, as it lacks FUNCTION_CALL steps for stubbed functions).

smowton committed:
Lazy goto model: hook for driver program to replace or stub functions

This provides a hook for the driver program to either replace a method outright, or to provide a
stub implementation when a function body cannot be provided by the language frontend (language_filest).

tautschnig commented:
Please check with #1272. Apparently there are N ways of marking something as do-not-display. Let's please not add approach N+1. The concept of "fallback functions" muddles the semantics of goto programs for the benefit of the user interface. At least that's my understanding of this matter...

smowton commented:
This isn't about hiding symbols or GOTO functions, it's about providing bodies for functions that we don't have source code for.

Example:

```java
class A {
  public int afield;

  static void f() {
    A a = Opaque.getA();
  }
}
```

If we don't have `Opaque.class` available, then `jbmc` currently replaces the callsite (using `convert_nondet`), so we get a program like:

```
f:
  A a;
  a = ALLOCATE(sizeof(A));
  a->afield = NONDET(int);
```

DeepTest on the other hand retains the call, but uses a custom derivative of `java_bytecode_languaget` that provides a body for the callee, so instead we get:

```
f:
  A a;
  a = Opaque.getA();

Opaque.getA():
  A ret = ALLOCATE(sizeof(A));
  ret->afield = NONDET(int);
  return ret;
```

That's more useful than JBMC's approach (I think JBMC should switch to using this method), as it (a) clarifies where the nondet init sequence came from (a call to stub `Opaque.getA()`), and (b) results in a trace containing `FUNCTION_CALL` and `FUNCTION_RETURN` steps for `Opaque.getA()`, which helps identify stub calls.

In order to do this, `lazy_goto_modelt` needs to know which `languaget` to ask to provide the body for `Opaque.getA`. The usual method of asking each `languaget` what methods it can provide doesn't work, as we actually don't have bytecode for `Opaque.getA`, but nontheless we can provide a definition if no other `languaget` is offering one. Hence we need some way to ask for the method body, even though it isn't a "claimed" method.

tautschnig commented:
Thanks a lot for the elaboration, but can we please have a comprehensive and agreed solution to address #1948, #1949, #2070, #1317, #1585, #1566 (and there might be more)? And of course right now, thanks to @hannes-steffenhagen-diffblue's work, we already have `--generate-function-body` in the public code base.

Maybe this PR is part of a bigger plan, but if so it would really help to sketch this out to make sure there isn't further fragmentation.

smowton commented:
Hmmm, okay, to cooperate better with these goto-instrument-oriented function generators / replacers we could instead ask the driver program what to do in this circumstance -- i.e., none of the language frontends could furnish a body for this front-end, so does the driver program want to have a go at providing a GOTO program somehow? Then it would use one of @hannes-steffenhagen-diffblue's passes, or for Java we'd do the nondet-object-graph thing. I can have a go at this solution next week.

smowton commented:
@tautschnig pushed an alternative implementation. I think this should satisfy everyone's needs (when using the lazy loading framework): it gives the driver program a chance to provide a body *before* `language_filest` (e.g. to replace bodies we *could* provide, as in @polgreen's block-function patch), but it can also opt to only provide missing functions (the stub use-case).

I'm checking with @peterschrammel whether we can open-source DeepTest's stub generation code; if so I plan to submit a followup PR to get jbmc using this mechanism to provide stubs.

tautschnig commented on src/goto-programs/lazy_goto_model.h:
```diff
@@ -110,6 +129,7 @@ class lazy_goto_modelt : public abstract_goto_modelt
   const goto_functionst::goto_functiont &get_goto_function(const irep_idt &id)
     override
   {
+
```
Unintended change?

tautschnig reviewed (commented):
I think this proposal is good, because it keeps the lazy behaviour contained rather than spreading it across multiple classes and APIs.

Yet I struggle to fully evaluate it (thus sticking with "Comment" rather than "Approve") as there are no regression tests (there may be tests exercising the new code paths already present in the test suite, but no test has observably different behaviour).

smowton commented:
If I get an ok to open a file of deeptest's code I will make jbmc a user of this code, then hopefully its existing tests will suffice to show this is working as intended.

tautschnig commented:
I, for one, am ok for the moment if you tell me that there are tests elsewhere that show this makes a difference and works as intended. Not great, but if you put that note in the commit message then you're committing to it ;-)

Elsewhere it was said that there are current issues in tests elsewhere, so you might, however, have to hold off on this anyway.

smowton commented:
This should be tested with Deeptest once it can build with mainline cbmc again.

smowton committed:
Add Java simple method stubbing pass

This generates simple stub bodies using the same method as convert_nondet, but creates
the stubs out-of-line rather than inline. That adds support for nondet constructors, which
jbmc previously ignored.

smowton committed:
JBMC: use simple method stubbing pass

This uses lazy_goto_modelt's support for driver programs supplying method replacements
to provide stub implementations of methods whose code is unavailable.

smowton commented:
@tautschnig @thk123 I've obtained permission from @peterschrammel to open a simple version of Deeptest's stub generator and added a commit to use it in jbmc.

The new stub generator, `java_bytecode/simple_method_stubbing.h/cpp`, has much reduced features compared to Deeptest's version, but is comparable to `convert_nondet`, which used to provide similar behaviour for jbmc.

If this works this will serve as a template for how Deeptest can integrate its own stub generator with lazy_goto_modelt.

smowton commented:
Test-gen bump: https://github.com/diffblue/test-gen/pull/1794

@tautschnig @peterschrammel @thk123 @mgudemann this is ready for review.

tautschnig commented on src/java_bytecode/simple_method_stubbing.h:
```diff
@@ -0,0 +1,34 @@
+#define CPROVER_JAVA_BYTECODE_SIMPLE_METHOD_STUBBING_H
+
+#include <util/symbol_table_base.h>
+#include <java_bytecode/java_bytecode_language.h>
```
Forward declarations would suffice.

tautschnig commented on src/java_bytecode/simple_method_stubbing.cpp:
```diff
@@ -0,0 +1,298 @@
+{
+  code_blockt new_instructions;
+  const code_typet &required_type = to_code_type(symbol.type);
+  namespacet ns(symbol_table);
```
Is that actually used in this method?

tautschnig commented on src/java_bytecode/simple_method_stubbing.cpp:
```diff
@@ -0,0 +1,298 @@
+#include <util/namespace.h>
+#include <util/prefix.h>
+#include <util/fresh_symbol.h>
+#include <util/invariant_utils.h>
```
Nit pick: lexicographic sorting.

tautschnig commented on src/java_bytecode/simple_method_stubbing.cpp:
```diff
@@ -0,0 +1,298 @@
+
+#include <util/std_expr.h>
+#include <util/std_code.h>
+#include <util/pointer_offset_size.h>
```
Is this needed/used?

tautschnig commented on src/java_bytecode/simple_method_stubbing.cpp:
```diff
@@ -0,0 +1,298 @@
+#include <util/std_code.h>
+#include <util/pointer_offset_size.h>
+#include <util/namespace.h>
+#include <util/prefix.h>
```
Seems unused.

tautschnig commented on src/java_bytecode/simple_method_stubbing.cpp:
```diff
@@ -0,0 +1,298 @@
+      synthesized_source_location,
+      ID_java,
+      symbol_table);
+    init_symbol.type = this_type;
```
Seems unnecessary?!

tautschnig commented on src/java_bytecode/simple_method_stubbing.cpp:
```diff
@@ -0,0 +1,298 @@
+        synthesized_source_location,
+        ID_java,
+        symbol_table);
+      to_return_symbol.type = required_return_type;
```
Seems unnecessary?!

tautschnig commented on src/java_bytecode/simple_method_stubbing.cpp:
```diff
@@ -0,0 +1,298 @@
+  if(sym.value.id()!=ID_nil)
+    return;
+  if(sym.type.id()!=ID_code)
+    return;
```
Nit pick: the use of `&&` would make this code more concise and would express what is the condition instead of saying when not to do this. Like this:
`if(!sym.is_type && sym.value.id() == ID_nil && sym.type.id() == ID_code) create_method_stub(...);`  (while moving the two `sym.name` to the top to clearly mark the exceptional case.

tautschnig commented on src/java_bytecode/simple_method_stubbing.cpp:
```diff
@@ -0,0 +1,298 @@
+void java_simple_method_stubst::check_method_stub(
+  const irep_idt &symname)
+{
+  const symbolt &sym=*symbol_table.lookup(symname);
```
This is a newly added file, so should be made clang-format correct.

tautschnig reviewed (approved):
Various, mostly stylistic notes.

smowton commented:
@tautschnig all changes applied.

tautschnig commented on src/java_bytecode/simple_method_stubbing.h:
```diff
@@ -0,0 +1,34 @@
+#include <util/irep.h>
+
+class message_handlert;
+class object_factory_parameterst;
```
Some compilers insist that this should be a `struct` and thus make the Travis build fail.

tautschnig commented:
`clang-format` has a lot to criticise - I'd suggest to make newly added code clang-format-clean.

smowton commented:
Fixed the compile bug. My local clang-format (LLVM 6.0) appears to disagree with the version running on Travis (LLVM 3.7), so I'm inclined to ignore them both.
